src/libcore/array.rs

```diff
@@ -264,6 +267,323 @@ macro_rules! array_impls {
+    {
+        #[inline]
+        fn eq(&self, other: &[A; N]) -> bool {
+            self[..] == other[..]
+        }
+        #[inline]
+        fn ne(&self, other: &[A; N]) -> bool {
+            self[..] != other[..]
+        }
+    }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<'a, 'b, A, B, const N: usize> PartialEq<&'b mut [B]> for [A; N]
+    where
+        A: PartialEq<B>,
+        [A; N]: LengthAtMost32,
+    {
+        #[inline]
+        fn eq(&self, other: &&'b mut [B]) -> bool {
+            self[..] == other[..]
+        }
+        #[inline]
+        fn ne(&self, other: &&'b mut [B]) -> bool {
+            self[..] != other[..]
+        }
+    }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<'a, 'b, A, B, const N: usize> PartialEq<[A; N]> for &'b mut [B]
+    where
+        B: PartialEq<A>,
+        [A; N]: LengthAtMost32,
+    {
+        #[inline]
+        fn eq(&self, other: &[A; N]) -> bool {
+            self[..] == other[..]
+        }
+        #[inline]
+        fn ne(&self, other: &[A; N]) -> bool {
+            self[..] != other[..]
+        }
+    }
+
+    // NOTE: some less important impls are omitted to reduce code bloat
+    // __impl_slice_eq2! { [A; $N], &'b [B; $N] }
+    // __impl_slice_eq2! { [A; $N], &'b mut [B; $N] }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<T: Eq, const N: usize> Eq for [T; N] where [T; N]: LengthAtMost32 {}
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<T: PartialOrd, const N: usize> PartialOrd for [T; N]
+    where
+        [T; N]: LengthAtMost32,
+    {
+        #[inline]
+        fn partial_cmp(&self, other: &[T; N]) -> Option<Ordering> {
+            PartialOrd::partial_cmp(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn lt(&self, other: &[T; N]) -> bool {
+            PartialOrd::lt(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn le(&self, other: &[T; N]) -> bool {
+            PartialOrd::le(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn ge(&self, other: &[T; N]) -> bool {
+            PartialOrd::ge(&&self[..], &&other[..])
+        }
+        #[inline]
+        fn gt(&self, other: &[T; N]) -> bool {
+            PartialOrd::gt(&&self[..], &&other[..])
+        }
+    }
+
+    #[stable(feature = "rust1", since = "1.0.0")]
+    impl<T: Ord, const N: usize> Ord for [T; N]
+    where
+        [T; N]: LengthAtMost32,
+    {
+        #[inline]
+        fn cmp(&self, other: &[T; N]) -> Ordering {
+            Ord::cmp(&&self[..], &&other[..])
+        }
+    }
+}
+
+/// Implemented for lengths where trait impls are allowed on arrays in core/std
+#[rustc_on_unimplemented(
+    message="arrays only have std trait implementations for lengths 0..=32",
+)]
+#[unstable(feature = "const_generic_impls_guard", issue = "0",
+    reason = "will never be stable, just a temporary step until const generics are stable")]
+#[cfg(not(bootstrap))]
+pub trait LengthAtMost32 {}
```

Comment:
Is it not possible to make this invisible outside `std`?

Reply:
The trait is unstable. Users can't do it unless they opt into instability.

Reply:
It's surprising that it's user-facing at all, even unstably. It's not a big problem though.

Reply:
I think it's good that it shows up in the `impl`s, so that they don't look like they work for all `N`.  And if it's going to be visible in the rustdoc there, I think it's better that the bound be clickable so people can see what it is.

Reply:
Hmm now I wonder whether the name `LengthAtMost32` makes sense or not, because from my non-native english understanding "at most N" means N inclusive. According to [cambridge dictionary](https://dictionary.cambridge.org/dictionary/english/most) "at most" is "used when saying what is the largest possible amount or number" so it matches my understanding. Shouldn't it be `LengthAtMost31` or `LengthLessThan32`?